RFC: Allow the `docker-proxy` to be overridden

djs55 opened the pull request:
This patch adds support for a new command-line argument
`--userland-proxy-bin=""` alongside the existing `--userland-proxy`
argument.

A replacement proxy binary is a convenient place to perform some
side-effect such as opening a tunnel to expose the port in some
other place over some other protocol. For example where docker is
running in a VM on a desktop, a replacement docker-proxy can use some
out-of-band mechanism to bind a port on the host, and send/receive
the data over some host<->VM channel such as [AF_VSOCK](https://lwn.net/Articles/646365/)
or [AF_HVSOCK](https://lkml.org/lkml/2015/7/28/404).

Other (less reliable) ways I've tried to solve the same problem include:
- watch docker API events and create/destroy tunnels on demand.
  Unfortunately errors are handled asynchronously and cannot cause
  the container start to fail.
- proxy the docker API itself and create/destroy tunnels during the
  API calls. This has the downside that it won't create tunnels when
  containers are automatically started by the daemon, and the
  docker API is larger (and faster moving) than the `docker-proxy` "API".

The new --userland-proxy-bin argument is only inspected if
--userland-proxy=true.

If --userland-proxy-bin="" (the default) then the existing built-in
docker-proxy will be used.

Edited to add: Note this PR doesn't include necessary changes like doc updates -- I thought it was best to get some feedback on the approach first, as we'd like to use it in Docker for Mac to replace our existing event-based implementation.

Signed-off-by: David Scott dave.scott@docker.com

djs55 committed:
Allow the `docker-proxy` to be overridden

This patch adds support for a new command-line argument
`--userland-proxy-bin=""` alongside the existing `--userland-proxy`
argument.

A replacement proxy binary is a convenient place to perform some
side-effect such as opening a tunnel to expose the port in some
other place over some other protocol. For example where docker is
running in a VM on a desktop, a replacement docker-proxy can use some
out-of-band mechanism to bind a port on the host, and send/receive
the data over some host<->VM channel such as AF_VSOCK.

Other (less reliable) ways I've tried to solve the same problem include:

- watch docker API events and create/destroy tunnels on demand.
  Unfortunately errors are handled asynchronously and cannot cause
  the container start to fail.

- proxy the docker API itself and create/destroy tunnels during the
  API calls. This has the downside that it won't create tunnels when
  containers are automatically started by the daemon, and the
  docker API is larger (and faster moving) than the `docker-proxy` "API".

The new --userland-proxy-bin argument is only inspected if
--userland-proxy=true.

If --userland-proxy-bin="" (the default) then the existing built-in
docker-proxy will be used.

Signed-off-by: David Scott <dave.scott@docker.com>

icecrime commented:
@thaJeztah Can we please track this janky failure? I'll rebuild:

```
07:14:45 FAIL: docker_cli_events_test.go:20: DockerSuite.TestEventsTimestampFormats
07:14:45 
07:14:45 docker_cli_events_test.go:41:
07:14:45     c.Assert(events, checker.HasLen, 2, check.Commentf("unexpected events, was expecting only 2 events tag/untag (since=%s, until=%s) out=%s", since, until, out))
07:14:45 ... obtained []string = []string{"2016-04-04T14:14:44.049490810Z image untag sha256:d9551b4026f0e2950ddb557cc640871710bf88476ca938b71499305647231b82 (name=sha256:d9551b4026f0e2950ddb557cc640871710bf88476ca938b71499305647231b82)"}
07:14:45 ... n int = 2
07:14:45 ... unexpected events, was expecting only 2 events tag/untag (since=1.199519662s, until=102.791278ms) out=2016-04-04T14:14:44.049490810Z image untag sha256:d9551b4026f0e2950ddb557cc640871710bf88476ca938b71499305647231b82 (name=sha256:d9551b4026f0e2950ddb557cc640871710bf88476ca938b71499305647231b82)
```

thaJeztah commented:
@icecrime it's already being tracked here; https://github.com/docker/docker/issues/21749, and pull request here https://github.com/docker/docker/pull/21750

thaJeztah commented:
/cc @mavenugo for libnetwork changes

thaJeztah commented on docs/reference/commandline/daemon.md:
```diff
@@ -67,6 +67,7 @@ weight = -1
       --tlsverify                            Use TLS and verify the remote
       --userns-remap="default"               Enable user namespace remapping
       --userland-proxy=true                  Use userland proxy for loopback traffic
+      --userland-proxy-bin=""                Use this userland proxy binary, if userland-proxy is set
```
Is this just the binary name, or the full path? If we expect a full path, should we call it `--userland-proxy-path` ?

On the other hand, the external credential stores use just the *name*, and requires the binary to be in `$PATH` (see https://github.com/docker/docker/pull/20107)

djs55 replied:
It's used as a `Cmd.Path` (i.e. an argv[0]) so perhaps it should be `--userland-proxy-path`?

aboch commented on vendor/src/github.com/docker/libnetwork/drivers/bridge/port_mapping.go:
```diff
@@ -14,7 +14,7 @@ var (
 )
 
-func (n *bridgeNetwork) allocatePorts(ep *bridgeEndpoint, reqDefBindIP net.IP, ulPxyEnabled bool) ([]types.PortBinding, error) {
+func (n *bridgeNetwork) allocatePorts(ep *bridgeEndpoint, reqDefBindIP net.IP, ulPxyEnabled bool, pxyBin string) ([]types.PortBinding, error) {
```
Instead of passing the custom proxy binary in all the functions up to `portmapper Map()`, could you instead add a new constructor in portmapper package which takes the custom proxy binary path ?
Or maybe change the existing `portmapper.New()` to `New(pxyBin string)`.

thaJeztah replied:
ping @djs55 could you have a look at @aboch's comment?

djs55 committed:
Rename --userland-proxy-bin to --userland-proxy-path

The value is used as a `Path` in `exec.Cmd` so it needs to be a path rather than
a simple filename.

Signed-off-by: David Scott <dave.scott@docker.com>

djs55 committed:
Add userland proxy path to portmapper.New() constructor

This avoids having to pass the argument through a large chain of
function calls.

Signed-off-by: David Scott <dave.scott@docker.com>

djs55 commented:
@thaJeztah thanks for the reminder! I've renamed `--userland-proxy-bin` to `--userland-proxy-path` as per your original comment, and have followed @aboch's suggestion of adding an argument to the `port mapper` `New()`: this has simplified the patch a lot.

I've uploaded my changes as additional changesets so you can see the changes, but can also squash them all together if you prefer.

cpuguy83 commented:
What about instead we split the proxy into a separate bin (since we already have multiple bins now anyway) and then do a path lookup? Users (or packagers) can replace the provided proxy bin with something else as needed.

justincormack commented:
@cpuguy83 yes that would work fine for our use case, and should improve the general memory usage too.

thaJeztah commented:
I like moving the proxy to a separate binary; however, even in that case, we probably need a way to specify the name or path of that binary; if users/packagers are using a different implementation, ideally it should also bear a different name (i.e. instead of `docker-proxy`, call it `myawesome-proxy`)

thaJeztah commented:
(it can default to automatically looking it up of course, just as we have with `docker-containerd`)

thaJeztah commented:
After discussing with @justincormack, we're closing this for now. There's going to be a new PR for moving the proxy out, and we'll discuss design for a possible flag on that PR
